api/infrastructure/parallelcluster-api.yaml

```diff
@@ -221,6 +221,93 @@ Resources:
       ImageTestsConfiguration:
         ImageTestsEnabled: false
 
+  EcrImageDeletionLambda:
+    Condition: DoNotUseCustomEcrImageUri
+    Type: AWS::Lambda::Function
+    Properties:
+      Code:
+        ZipFile: |
+          import cfnresponse
+          import boto3
+
+          def get_tag_from_image_arn(image_arn):
+              """ Returns the Docker image tag given an ImageBuilder image ARN
+
+              Arn format: arn:aws:imagebuilder:${AWS::Region}:${AWS::AccountId}:image/.../${Version}/${BuildNumber}
+              Tag format: ${Version}-${BuildNumber}
+              """
+              parts = image_arn.split('/')
+              version = parts[-2]
+              build_number = parts[-1]
+              return '{}-{}'.format(version, build_number)
+
+          def handler(event, context):
```

Comment:
Regarding error handling:
1. I agree with you that we can better handle the case where the ECR image has been deleted already without failing.
2. We should probably handle failures and send a failure response to CFN otherwise CFN will wait until the Lambda times-out
3. when sending a failure event to CFN can we also send the reason of the failure (e.g. the failure from the boto3 call)
4. Let's male sure to have the proper boto3 retry in place (the default might be fine but let's double-check it) cause once the lambda fails this cannot be retried by CFN.

Reply:
Regarding the retries, the client does 3 retries by default. Regrading the code, I have wrapped the handler in a try/catch and failed fast by returning a failure response and putting the exception in the status reason. I checked what was the Boto3 version, and verified that there is an exception that we might want to catch and "ignore" (i.e. succeeding anyway), namely [ECR.Client.exceptions.RepositoryNotFoundException](https://boto3.amazonaws.com/v1/documentation/api/1.17.42/reference/services/ecr.html#ECR.Client.batch_delete_image). Not sure about this, my current reasoning is:
1. In general if the customer deletes resources created by the stack it is correct to give an error, so we might want to fail deletion if we catch this exception
2. In the case of trying to remove an image that does not exist in the repository, the API does not give an exception, so we don't fail deletion: this is consistent with the current PC behavior elsewere, plus the Docker image is not a resource in the CFN template (it is "created indirectly" through the definition of the ImageBuilder image, which is the [AWS::ImageBuilder::Image
](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-imagebuilder-image.html) resource in the CFN template)

Reply:
I think it's resonable never to fail deletion if resources are not present